velox/expression/Expr.cpp

```diff
@@ -572,8 +572,10 @@ void Expr::eval(
        topLevel ? (void*)&exprExceptionContext : this});
 
   if (!rows.hasSelections()) {
-    // empty input, return an empty vector of the right type
-    result = BaseVector::createNullConstant(type(), 0, context.pool());
+    if (!result) {
```

Comment:
I wonder if it would be better to throw if 'rows' is empty.

```
VELOX_CHECK(rows.hasSelections());
```

Reply:
should we also add similar handling in evalFlatNoNulls() at L563?

Reply:
@mbasmanova I tried this but realized that if we simply check rows.hasSelections() in Expr::eval, then we need to have `if (!rows.hasSelection()) { if (!result) { result = BaseVector::createNullConstant(...);} return; }` at all callsites of Expr::eval. In this case, I'd prefer we centralize this handling in Expr::eval only. What do you think?

Reply:
> I wonder if it would be better to throw if 'rows' is empty.
> 
> ```
> VELOX_CHECK(rows.hasSelections());
> ```

Hi @mbasmanova, as we discussed offline, ideally we'd like to make an expression evaluation terminate early if there is no selected rows and check that Expr::eval never gets empty selectivity vector. To achieve this, we need to ensure two things:
1. ExprSet::eval never pass empty selectivity vector to Expr::eval (same for ExprSetSimplified).
2. Any place where selected rows can be reduced during evaluation needs a check that remaining rows is not empty. These places are in SwitchExpr, CoalesceExpr, ConjunctExpr, Expr::evalArgsDefaultNulls(), and Expr::evalArgsWithNulls().

I've checked that these places except SwitchExpr all have proper checks for selectivity vector being non-empty before calling Expr:eval, and this PR adds the check in SwitchExpr. 

However, since this PR adds additional checks that could throw at runtime, I wonder how we should test it to ensure it doesn't affect users. For example, I noticed that the existing unit test ExprTest.testEmptyVectors fails after this change.

One possible way for testing is to add an `always_throw(bool) -> bool` function to expression fuzzer so that some corner cases can be triggered when it's used as a switch condition evaluated in try(). Another thing we can do is to run e2e tests for internal customers. What do you think?

Reply:
Hi @mbasmanova, I've tried adding a check that makes Expr::eval and Expr::evalSimplified throw when they receive an empty selectivity vector, and it passed all existing unit tests. However, after discussing with @laithsakka, we think it would be difficult to prove that there won't be such a corner case where Expr::eval is called with an empty selectivity vector during an evaluation. Developers would also have to remember this restriction when they make changes to the evaluator. So I'm changing it back to let Expr::eval and Expr::evalSimplified accept empty selectivity vector to avoid introducing potential bugs.